File: vgcbench/choices.py
from typing import Literal, TypedDict


class MenuItem(TypedDict):
    label: str
    part: str
    kind: Literal["move", "switch", "team", "pass"]


SlotMenu = list[MenuItem]

_TARGETS = {"normal", "any", "adjacentFoe"}
_SPREAD = {
    "allySide",
    "foeSide",
    "all",
    "allAdjacent",
    "allAdjacentFoes",
    "allies",
}


def _name(pokemon: dict) -> str:
    ident = pokemon.get("ident", "")
    return ident.split(": ", 1)[-1] or pokemon.get("details", "Pokémon").split(",", 1)[0]


def _switches(request: dict, reviving: bool = False) -> SlotMenu:
    menu: SlotMenu = []
    for index, pokemon in enumerate(request.get("side", {}).get("pokemon", []), 1):
        fainted = pokemon.get("condition", "").endswith(" fnt")
        if pokemon.get("active") or fainted != reviving:
            continue
        menu.append({"label": f"Switch to {_name(pokemon)}", "part": f"switch {index}", "kind": "switch"})
    return menu


def _target_label(side: str, number: int, names: dict | None) -> str:
    species = (names or {}).get(side, {}).get(number)
    return f" -> {side} {number}" + (f" ({species})" if species else "")


def _move_items(
    move: dict,
    slot: int,
    active_count: int,
    has_ally: bool,
    mega: bool,
    names: dict | None = None,
) -> SlotMenu:
    move_slot = move["slot"]
    name = move.get("move", f"Move {move_slot}")
    target = move.get("target")
    targets: list[tuple[str, str]] = [("", "")]
    if active_count > 1 and target in _TARGETS:
        targets = [(f" +{number}", _target_label("foe", number, names)) for number in (1, 2)]
        if target in {"normal", "any"} and has_ally:
            ally = 2 if slot == 1 else 1
            targets.append((f" -{ally}", _target_label("ally", ally, names)))
    elif active_count > 1 and target == "adjacentAlly":
        ally = 2 if slot == 1 else 1
        targets = [(f" -{ally}", _target_label("ally", ally, names))]
    elif active_count > 1 and target == "adjacentAllyOrSelf":
        targets = []
        for number in (1, 2):
            label = " -> itself" if number == slot else _target_label("ally", number, names)
            targets.append((f" -{number}", label))

    menu: SlotMenu = []
    spread = " (spread)" if target in _SPREAD else ""
    for target_part, target_label in targets:
        part = f"move {move_slot}{target_part}"
        label = f"{name}{spread}{target_label}"
        menu.append({"label": label, "part": part, "kind": "move"})
        if mega:
            menu.append({"label": f"{label} + Mega Evolve", "part": f"{part} mega", "kind": "move"})
    return menu


def build_menus(request: dict, names: dict | None = None) -> list[SlotMenu]:
    side = request.get("side", {})
    pokemon = side.get("pokemon", [])
    if request.get("wait"):
        return []
    if request.get("teamPreview"):
        count = request.get("maxChosenTeamSize") or len(pokemon)
        base = [
            {"label": f"Pick {_name(mon)}", "part": str(index), "kind": "team"}
            for index, mon in enumerate(pokemon, 1)
        ]
        return [[item.copy() for item in base] for _ in range(count)]
    if "forceSwitch" in request:
        menus: list[SlotMenu] = []
        active_pokemon = [mon for mon in pokemon if mon.get("active")]
        for slot, forced in enumerate(request["forceSwitch"]):
            if not forced:
                menus.append([{"label": "Pass", "part": "pass", "kind": "pass"}])
                continue
            current = active_pokemon[slot] if slot < len(active_pokemon) else {}
            menu = _switches(request, bool(current.get("reviving")))
            menus.append(menu or [{"label": "Pass", "part": "pass", "kind": "pass"}])
        forced = [index for index, value in enumerate(request["forceSwitch"]) if value]
        switch_targets = {item["part"] for menu in menus for item in menu if item["kind"] == "switch"}
        if len(switch_targets) < len(forced):
            for index in forced:
                if not any(item["kind"] == "pass" for item in menus[index]):
                    menus[index].append({"label": "Pass", "part": "pass", "kind": "pass"})
        return menus
    if "active" in request:
        menus = []
        active_pokemon = [mon for mon in pokemon if mon.get("active")]
        active_count = len(request["active"])
        for slot, active in enumerate(request["active"], 1):
            current = active_pokemon[slot - 1] if slot <= len(active_pokemon) else {}
            if active is None or current.get("commanding") or current.get("condition", "").endswith(" fnt"):
                menus.append([{"label": "Pass", "part": "pass", "kind": "pass"}])
                continue
            menu: SlotMenu = []
            enabled_moves = [move for move in active.get("moves", []) if not move.get("disabled")]
            for move_slot, move in enumerate(active.get("moves", []), 1):
                if move.get("disabled"):
                    continue
                menu.extend(
                    _move_items(
                        {**move, "slot": move_slot},
                        slot,
                        active_count,
                        active_count > 1 and request["active"][1 if slot == 1 else 0] is not None,
                        bool(active.get("canMegaEvo")),
                        names,
                    )
                )
            if not enabled_moves and active.get("moves"):
                menu.append({"label": "Struggle", "part": "move 1", "kind": "move"})
            if not active.get("trapped"):
                menu.extend(_switches(request))
            menus.append(menu or [{"label": "Pass", "part": "pass", "kind": "pass"}])
        return menus
    return [[{"label": "Pass", "part": "pass", "kind": "pass"}]]


def compose(parts: list[str]) -> str:
    return ", ".join(parts)

File: src/choices.ts
import { PROTECT_MOVES } from './state.js';
import type { BattleRequest, JsonObject } from './types.js';

import { afterColon, asRecords, text } from './value.js';

type MenuKind = 'move' | 'switch' | 'team' | 'pass' | 'forfeit';
interface MenuItem {
  label: string;
  part: string;
  kind: MenuKind;
}
export type SlotMenu = MenuItem[];
export type TargetNames = Record<'foe' | 'ally', Record<number, string>>;

export interface MenuHints {
  names?: TargetNames;
  /** Per active slot: whether consecutive Protect odds are reduced. */
  protectReduced?: Record<number, boolean>;
  moveAnnotation?: (slot: number, move: string, targetSide: 'foe' | 'ally', targetNumber: number) => string | undefined;
}

const SELECTED_TARGETS = new Set(['normal', 'any', 'adjacentFoe']);
const SPREAD_TARGETS = new Set(['allySide', 'foeSide', 'all', 'allAdjacent', 'allAdjacentFoes', 'allies']);

function pokemonSpecies(pokemon: JsonObject): string {
  const fromDetails = text(pokemon.details).split(',', 1)[0]?.trim();
  if (fromDetails) return fromDetails;
  return afterColon(text(pokemon.ident)) || 'Pokémon';
}

function switches(request: BattleRequest, reviving = false): SlotMenu {
  const menu: SlotMenu = [];
  for (const [index, pokemon] of (request.side?.pokemon ?? []).entries()) {
    const fainted = text(pokemon.condition).endsWith(' fnt');
    if ((!reviving && pokemon.active) || fainted !== reviving) continue;
    menu.push({ label: `Switch to ${pokemonSpecies(pokemon)}`, part: `switch ${index + 1}`, kind: 'switch' });
  }
  return menu;
}

function spreadLabel(target: string, slot: number, names?: TargetNames): string {
  if (target === 'allAdjacentFoes' || target === 'foeSide') return ' (both foes)';
  if (target === 'allies' || target === 'allySide') {
    const ally = slot === 1 ? 2 : 1;
    const species = names?.ally[ally];
    return species ? ` (your side, including ally ${species})` : ' (your side)';
  }
  if (target === 'allAdjacent' || target === 'all') {
    const ally = slot === 1 ? 2 : 1;
    const species = names?.ally[ally];
    return species ? ` (all adjacent, including ally ${species})` : ' (all adjacent, including ally)';
  }
  return ' (spread)';
}

function moveItems(
  move: JsonObject,
  moveSlot: number,
  slot: number,
  activeCount: number,
  hasAlly: boolean,
  mega: boolean,
  hints?: MenuHints,
): SlotMenu {
  const target = text(move.target);
  const names = hints?.names;
  const targetLabel = (side: 'foe' | 'ally', number: number) => {
    const species = names?.[side][number];
    return ` -> ${side} ${number}${species ? ` (${species})` : ''}`;
  };
  let targets: Array<[string, string]> = [['', '']];
  if (activeCount > 1 && SELECTED_TARGETS.has(target)) {
    targets = [1, 2].map((number) => [` +${number}`, targetLabel('foe', number)]);
    if ((target === 'normal' || target === 'any') && hasAlly) {
      const ally = slot === 1 ? 2 : 1;
      targets.push([` -${ally}`, targetLabel('ally', ally)]);
    }
  } else if (activeCount > 1 && target === 'adjacentAlly') {
    const ally = slot === 1 ? 2 : 1;
    targets = hasAlly ? [[` -${ally}`, targetLabel('ally', ally)]] : [];
  } else if (activeCount > 1 && target === 'adjacentAllyOrSelf') {
    const ally = slot === 1 ? 2 : 1;
    targets = [[` -${slot}`, ' -> itself']];
    if (hasAlly) targets.push([` -${ally}`, targetLabel('ally', ally)]);
  }

  const name = text(move.move, `Move ${moveSlot}`);
  const protectHint =
    PROTECT_MOVES.has(name.toLowerCase().replace(/[^a-z0-9]+/g, '')) && hints?.protectReduced?.[slot]
      ? ' [success rate reduced: Protected last turn]'
      : '';
  const spread = SPREAD_TARGETS.has(target) ? spreadLabel(target, slot, names) : '';
  return targets.flatMap(([targetPart, label]) => {
    const part = `move ${moveSlot}${targetPart}`;
    const selected = /^ ([+-])([12])$/.exec(targetPart);
    const annotation = selected
      ? hints?.moveAnnotation?.(slot, name, selected[1] === '+' ? 'foe' : 'ally', Number(selected[2]))
      : undefined;
    const warning = annotation ? ` [${annotation}]` : '';
    const item: MenuItem = { label: `${name}${spread}${protectHint}${label}${warning}`, part, kind: 'move' };
    return mega ? [item, { label: `${item.label} + Mega Evolve`, part: `${part} mega`, kind: 'move' }] : [item];
  });
}

export function buildMenus(request: BattleRequest, hints?: MenuHints): SlotMenu[] {
  const pokemon = request.side?.pokemon ?? [];
  if (request.wait) return [];
  if (request.teamPreview) {
    const count = request.maxChosenTeamSize || pokemon.length;
    const base = pokemon.map(
      (mon, index): MenuItem => ({ label: `Pick ${pokemonSpecies(mon)}`, part: String(index + 1), kind: 'team' }),
    );
    return Array.from({ length: count }, () => base.slice());
  }
  if (request.forceSwitch) {
    const active = pokemon.filter((mon) => mon.active);
    const menus = request.forceSwitch.map((forced, slot): SlotMenu => {
      if (!forced) return [{ label: 'Pass', part: 'pass', kind: 'pass' }];
      const menu = switches(request, Boolean(active[slot]?.reviving));
      return menu.length ? menu : [{ label: 'Pass', part: 'pass', kind: 'pass' }];
    });
    const forced = request.forceSwitch.flatMap((value, index) => (value ? [index] : []));
    const targets = new Set(
      menus.flatMap((menu) => menu.filter((item) => item.kind === 'switch').map((item) => item.part)),
    );
    if (targets.size < forced.length) {
      for (const index of forced) {
        const menu = menus[index]!;
        if (!menu.some((item) => item.kind === 'pass')) menu.push({ label: 'Pass', part: 'pass', kind: 'pass' });
      }
    }
    if (menus.some((menu) => menu.some((item) => item.kind === 'switch')))
      menus[0]!.push({ label: 'Forfeit the game (concede the loss)', part: 'forfeit', kind: 'forfeit' });
    return menus;
  }
  if (request.active) {
    const activePokemon = pokemon.filter((mon) => mon.active);
    const menus = request.active.map((active, slotIndex): SlotMenu => {
      const slot = slotIndex + 1;
      const current = activePokemon[slotIndex] ?? {};
      if (!active || current.commanding || text(current.condition).endsWith(' fnt')) {
        return [{ label: 'Pass', part: 'pass', kind: 'pass' }];
      }
      const moves = asRecords(active.moves);
      const allyIndex = slot === 1 ? 1 : 0;
      const ally = activePokemon[allyIndex];
      const hasAlly =
        request.active!.length > 1 &&
        request.active![allyIndex] !== null &&
        ally !== undefined &&
        !ally.commanding &&
        !text(ally.condition).endsWith(' fnt');
      const menu = moves.flatMap((move, index) =>
        move.disabled
          ? []
          : moveItems(move, index + 1, slot, request.active!.length, hasAlly, Boolean(active.canMegaEvo), hints),
      );
      if (moves.length && moves.every((move) => move.disabled))
        menu.push({ label: 'Struggle', part: 'move 1', kind: 'move' });
      if (!active.trapped) menu.push(...switches(request));
      return menu.length ? menu : [{ label: 'Pass', part: 'pass', kind: 'pass' }];
    });
    if (menus.some((menu) => menu.some((item) => item.kind !== 'pass')))
      menus[0]!.push({ label: 'Forfeit the game (concede the loss)', part: 'forfeit', kind: 'forfeit' });
    return menus;
  }
  return [[{ label: 'Pass', part: 'pass', kind: 'pass' }]];
}
